src/main/java/org/opensearch/flowframework/util/ParseUtils.java

```diff
@@ -229,6 +247,218 @@ public static User getUserContext(Client client) {
         return User.parse(userStr);
     }
 
+    /**
+     * Add user backend roles filter to search source builder=
+     * @param user the user
+     * @param searchSourceBuilder search builder
+     * @return search builder with filter added
+     */
+    public static SearchSourceBuilder addUserBackendRolesFilter(User user, SearchSourceBuilder searchSourceBuilder) {
+        if (user == null) {
+            return searchSourceBuilder;
+        }
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        String userFieldName = "user";
+        String userBackendRoleFieldName = "user.backend_roles.keyword";
+        List<String> backendRoles = user.getBackendRoles() != null ? user.getBackendRoles() : ImmutableList.of();
+        // For normal case, user should have backend roles.
+        TermsQueryBuilder userRolesFilterQuery = QueryBuilders.termsQuery(userBackendRoleFieldName, backendRoles);
+        NestedQueryBuilder nestedQueryBuilder = new NestedQueryBuilder(userFieldName, userRolesFilterQuery, ScoreMode.None);
+        boolQueryBuilder.must(nestedQueryBuilder);
+        QueryBuilder query = searchSourceBuilder.query();
+        if (query == null) {
+            searchSourceBuilder.query(boolQueryBuilder);
+        } else if (query instanceof BoolQueryBuilder) {
+            ((BoolQueryBuilder) query).filter(boolQueryBuilder);
+        } else {
+            // e.g., wild card query
+            throw new FlowFrameworkException("Search API does not support queries other than BoolQuery", RestStatus.BAD_REQUEST);
```

Comment:
Why not?  We can wrap an existing query of any type in a BoolQuery `must()` and then add a filter.  See [example](https://github.com/opensearch-project/ml-commons/blob/9f4d2ce47c1f53b2003733b91f37a1246182e1ff/common/src/main/java/org/opensearch/sdk/client/LocalClusterIndicesClient.java#L227-L230)

Reply:
Yes, this looks better with no hits
```json
{
    "took": 11,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 0,
            "relation": "eq"
        },
        "max_score": null,
        "hits": []
    }
}
```e